File: 6 kyu/Campfire building/solution.py
from math import sqrt

def is_constructable(area):
    '''
    Is "area" the sum of two perfect squares?
    '''
    return any(
        sqrt(area - num ** 2).is_integer()
        for num in range(int(sqrt(area)) + 1)
    )
__________________________________
def is_constructable(a):
    i = 2
    while (i * i <= a):
        c = 0
        if (a % i == 0):
            while (a % i == 0):
                a = int(a / i)
                c += 1
            if (i % 4 == 3 and c % 2 != 0):
                return False
        i += 1
    return a % 4 != 3
__________________________________
import math
def is_constructable(a):
    r = math.sqrt(a)
    x = math.floor(r)
    y = 0
    while x >= y:
        if a == (x*x)+(y*y):
            return True
        if a < (x*x)+(y*y):
            x -= 1
            y += 1
        else:
            y += 1
    return False
__________________________________
def is_constructable(a):
    i=0
    while True:
        if i**2>a:
            return False
        if (a-i**2)**0.5-int((a-i**2)**0.5)==0:
            return True
        i+=1
__________________________________
def is_constructable(a):
    if a ** (1/2) % 1 == 0 : return True
    for i in range(1, int((a/2)**(1/2)) + 1) :
        if ((a - (i ** 2)) ** (1 / 2)) % 1 == 0 : return True
    return False
__________________________________
def is_constructable(a):
    if a ==1:
        return True
    s = dict()
    for i in range(a):
        if  i*i >a:
            break
        s[i*i]=1
        if (a-i*i) in s.keys():
            return True
    return False
__________________________________
import math

def is_constructable(area):
    """ The problem is recast as finding Pythagorean triples (also perfect
        squares by allowing for a 'zero side'), as the area is equivalent
        to c^2 = a^2 + b^2. The below logic verifies that c^2 - a^2 = b^2,
        where b is an integer, by checking that the float square root is
        equal to the integer square root. """
    max_side = math.isqrt(area) + 1
    for i in range (0, max_side):
        if math.isqrt(area - i*i) == math.sqrt(area - i*i):
            return True
    return False
__________________________________
def is_constructable(a):
    for n in range(int(a**0.5) + 1):
        if ((a - n**2)**0.5).is_integer():
            return True
    return False
__________________________________
def is_constructable(a):
    if ((int(a ** (1/2))) ** 2) == a:
        return (True)
    counter = 0
    while counter * counter < a:
        counter += 1
        k = counter * counter
        y = a - k
        try:
            j = ((int(y ** (1/2))) ** 2)
            if j == y:
                return True
        except:
            pass
    return False

File: 6 kyu/Campfire building/solution.c
#include <stdbool.h>
#include <math.h>

bool is_constructable(long a)
{
  long n = floor(sqrt(a));
  // Try to find a y, such that x² + y² = a
  // (This is an application of Pythagoras)
  for (int x = n; x >= 0; x--) {
    double y = sqrt(a - pow(x, 2));
    if (y == floor(y))
      return true;
  }
  return false;
}
__________________________________
#include <stdbool.h>
#include <math.h>

long sqrt_long(long x)
{
    return (long)(sqrt(x) + 1e-6);
}

bool is_perfect_square(long x)
{
    long t = sqrt_long(x);
    return t * t == x;
}

bool is_constructable(long a)
{
    long m = sqrt_long(a);
    for (long x = 0; x <= m; ++x) {
        if (is_perfect_square(a - x * x)) return true;
    }
    return false;
}
__________________________________
#include <math.h>
#include <stdbool.h>

bool is_constructable(long area) {
    long max_side = sqrt(area);
    for (long i = 0; i <= max_side; i++) {
        double side = sqrt(area - i * i);
        if (side == (long) side) {
            return true;
        }
    }
    return false;
}
__________________________________
#include <math.h>
#include <stdbool.h>

bool is_constructable(long area) {
    // Is "area" the sum of two perfect squares?
    long side = sqrt(area);
    for (long num = 0; num <= side; num++) {
        double temp = sqrt(area - num * num);
        if (temp == (long) temp) {
            return true;
        }
    }
    return false;
}
__________________________________
#include <stdbool.h>
#include <math.h>

bool is_constructable(long a)
{
  long sq = sqrt(a);
  if(sq*sq == a) return true;
  for(long i=1;i<=sq;i++) 
    for(long j=sq;j>0;j--) 
      if(a == i*i + j*j) return true;
  return false;
}
